feat: add sanity swap checks after handshake

sangaman opened the pull request:
Closes #676.

This is a major commit that changes the procedure for accepting orders from peers to ensure that we can send and receive payments for both currencies that make up the order's trading pair. This introduces the concept of "sanity swaps," which are atomic swaps of 1 satoshi for 1 satoshi within a single currency/chain. The flow for atomic swaps is as follows:

1. A connection is established with a peer.
2. The handshake is performed successfully.
3. A `SanitySwapPacket` will be sent to the peer to inform them of a sanity swap attempt for each currency the appears in the peer's trading pairs advertised in the `nodeState`.
4. We send a 1 satoshi payment to the peer for each sanity swap.
5. Upon receiving our 1 satoshi payment, the peer attempts to learn the preimage by sending a 1 satoshi payment to us.
6. We release the preimage that we created, effectively settling the swap.
7. After all sanity swaps resolve, we check which advertised pairs had successful swaps for both the base and quote currency.
8. These pairs are marked "active" and we send a `GetOrdersPacket` for the newly activated pairs.

Other notable changes include:

- Tracking "active" pairs on a per-peer basis. "Active" pairs are those that have been advertised by a peer's node state, which we have been able to successfully perform a sanity swap for, and which have not encountered unexpected payment errors from our perspective.
- Not requesting all orders via a `GetOrdersPacket` automatically upon a successful connection and handshake.
- Ignoring order packets we receive for trading pairs that are not active.
- The `ListPeers` rpc call only returns active trading pairs.
- Sanity swaps can be disabled via the `nosanitychecks` config option.

BREAKING CHANGE: New SanitySwapPacket.

erkarl commented:
Nice! I'm excited to get this feature in master as it will improve the likelihood of having real swaps fail. I went through the code once and got a pretty good overview of the implementation details - looks good on a high level. 👍 

This PR is quite big - I will do a more thorough/low-level review of the code later. Will also allocate time test this manually.

erkarl commented:
Suggestions based on the weekly call:
- make the feature optional
- implement retry logic

kilrau commented:
Can we get this one in?

sangaman commented:
@erkarl @moshababo I've rebased and updated this PR so I think it's ready for review. Tests are currently failing due to a bug that I think came from #863 that causes the simulation tests to fail if `nosanitychecks` is disabled (which I want to do so that these sanity swaps get tested). I'm looking into fixing that in a separate PR.

sangaman commented:
Will rebase on top of #901 when it's merged which should fix the tests.

sangaman commented:
Rebased and passing the tests with sanity swaps enabled. @erkarl @moshababo

sangaman commented on lib/orderbook/OrderBook.ts:
```diff
@@ -645,6 +672,73 @@ class OrderBook extends EventEmitter {
+
+    if (this.swaps) {
+      // Set a time limit for all sanity swaps to complete.
+      const sanitySwapTimeout = setTimeoutPromise(15000, false);
```
Yes and will also add a few code comments around this block.
